compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -286,6 +288,86 @@ pub enum Visibility<Id = LocalDefId> {
     Restricted(Id),
 }
 
+#[derive(Clone, Debug, PartialEq, Eq, Copy, Hash, Encodable, Decodable, HashStable)]
+pub enum Restriction<Id = DefId> {
+    /// The restriction does not affect the item.
+    Unrestricted,
+    /// The restriction only applies outside of this path.
+    Restricted(Id, Span),
+}
+
+impl<Id: Into<DefId> + Copy> Restriction<Id> {
+    /// Returns `true` if this restriction applies in the given module. This
+    /// means the behavior is _not_ allowed.
+    pub fn is_restricted_in(self, module: DefId, tcx: TyCtxt<'_>) -> bool {
```

Comment:
I would suggest inverting this function, making it into a `is_allowed_in`, imo this would be a bit more easy to understand.